### PR TITLE
Change icu_pattern PlaceholderValueProvider and parts behavior

### PR DESCRIPTION
**Pull request by sffc**
See discussion in #5897

### CODE REVIEW

File: components/pattern/src/common.rs [hunk 1]
```diff
@@ -110,56 +110,137 @@ pub trait PatternBackend: crate::private::Sealed + 'static + core::fmt::Debug {
...
+/// [`Part`]: writeable::Part
 pub trait PlaceholderValueProvider<K> {
     type Error;
 
     type W<'a>: TryWriteable<Error = Self::Error>
     where
         Self: 'a;
 
-    const LITERAL_PART: Part;
+    type L<'a, 'l>: Writeable
```
Comment (Manishearth) [inline components/pattern/src/common.rs:211]: nit: type params need docs.

**Review by Manishearth — APPROVED** [review]
Not clear that this is necessary if we're splitting interpolate, but it doesn't hurt to have this flexibility, it makes sense that if placeholders can be intercepted, then literals can too.

**Comment by sffc** [comment]
> Not clear that this is necessary if we're splitting interpolate, but it doesn't hurt to have this flexibility, it makes sense that if placeholders can be intercepted, then literals can too.

We won't need to split interpolate with this.

**Comment by Manishearth** [comment]
Yeah, I prefer splitting interpolate, but this is fine too. I'd like more docs so it's clear which choice to make and how. Not urgent.

**Comment by sffc** [comment]
The new trait basically means "the pattern crate does not add its own parts unless you work hard" which seems fine. The pattern crate does not generally contribute semantic meaning, and it should be able to be swapped out with some other crate without affecting formatToParts behavior. In the one known case where you need it to add parts (range formatting), we have a power-user API to let you do it.

File: components/pattern/src/common.rs [hunk 2]
```diff
@@ -110,56 +110,137 @@ pub trait PatternBackend: crate::private::Sealed + 'static + core::fmt::Debug {
...
+/// };
+///
+/// impl PlaceholderValueProvider<DoublePlaceholderKey> for ValuesWithParts<'_> {
+///     type Error = core::convert::Infallible;
+///
+///     type W<'a> = WriteableAsTryWriteableInfallible<WithPart<&'a str>>
+///     where
+///         Self: 'a;
+///
+///     type L<'a, 'l> = WithPart<&'l str>
```
Comment (sffc) [inline components/pattern/src/common.rs:162]: Also, I really like how this solution composes cleanly with the Writeable adapters I added a few months ago like `WithPart`.

**Pull request merged**